src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,198 @@
+        {
+            m_current_index = Shape(m_input_shape.size(), 0UL);
+            const auto values_count = shape_size(m_input_shape);
+
+            const T zero_value = T{0};
+            for (size_t i = 0; i < values_count; ++i)
+            {
+                if (values[i] != zero_value)
+                {
+                    add_to_results(m_current_index);
+                }
+
+                // don't generate the next index when the last value has been processed
+                if (i < values_count - 1)
+                {
+                    next_index();
+                }
+            }
+
+            return m_results;
+        }
+
+    private:
+        /// \brief Adds single dimensions of an index into the matching element of the results
+        void add_to_results(const Shape& index)
+        {
+            for (size_t dim = 0; dim < index.size(); ++dim)
+            {
+                m_results.at(dim).push_back(index[dim]);
+            }
+        }
+
+        // Generates an index pointing to the next element in the flattened tensor
+        // It behaves similar to flipping bits when incrementing a binary number
+        void next_index()
+        {
+            for (size_t dim = m_current_index.size() - 1; dim >= 0; --dim)
+            {
+                auto& dim_value = m_current_index.at(dim);
+                if (dim_value + 1 == m_input_shape[dim])
+                {
+                    dim_value = 0;
+                }
+                else
+                {
+                    ++dim_value;
+                    return;
+                }
+            }
+        }
+
+    private:
+        const Shape m_input_shape;
+        Results_t m_results;
+        Shape m_current_index;
+    };
+}
+
+template <typename T>
+static shared_ptr<op::Constant> fold_constant_non_zero(const shared_ptr<op::Constant>& data)
+{
+    const auto input_shape = data->get_shape();
+    const auto* input_values = data->get_data_ptr<T>();
+    if (ngraph::is_scalar(input_shape))
+    {
+        const auto scalar_value = input_values[0];
+
+        NGRAPH_CHECK(scalar_value != T{0},
+                     "It's not possible to constant fold a NonZero op for a scalar equal to zero.");
+
+        // return 0(the only index) if the data input contains a scalar different than zero
+        return op::Constant::create(element::i64, Shape{1, 1}, {0});
+    }
+    else if (is_vector(input_shape))
+    {
+        const auto input_values_count = shape_size(input_shape);
+        std::vector<int64_t> indices;
+        indices.reserve(input_values_count);
+
+        const T zero_value = T{0};
+        for (size_t i = 0; i < input_values_count; ++i)
+        {
+            if (input_values[i] != zero_value)
+            {
+                indices.push_back(i);
+            }
+        }
+
+        indices.shrink_to_fit();
+        return op::Constant::create(element::i64, Shape{1, indices.size()}, indices);
+    }
```

Comment:
Have you tried 1D data with you helper structure`NonZeroElements`? I suppose it might work.

Reply:
Yes, it does work but the point of this specialization for vectors is to have an even faster implementation. In this `else if` branch there's less allocation, less copying and less processing of indices.